Narration

jwonderly opened the pull request:
Adds support for narration of calls to <code>umsg()</code> for slideshows, making use of the [SpeechSynthesis](https://developer.mozilla.org/en-US/docs/Web/API/SpeechSynthesis) functionality of the user's browser.

Narration is disabled by default. It can be enabled globally by setting <code>JSAV_OPTIONS.narrationEnabled = true</code>, and it can be enabled/disabled for an individual slideshow.

Enabling narration will add a button to the left of the settings button that users can click to toggle narration for the slideshow.

When a user clicks the narration button, an event is logged. If clicking the button turns on narration, then <code>jsav-narration-on</code> is logged. If clicking the button turns off narration, then <code>jsav-narration-off</code> is logged.

The vast majority of the work for this feature was done by @ehsanelgendi.

vkaravir commented:
Hi @jwonderly,
Thank you for working on this! And apologies for taking some time to review.

Overall looks good, but there's one thing I hope you could change before I merge this into master. Currently this will break in browsers that don't support the Speech Synthesis API. I'd like it to check for the support for `speechSynthesis`and SpeechSynthesisUtterance` before using those features. While JSAV doesn't have a list of officially supported browsers, there's no need for this feature to break older browsers. Could you change the code to check for these?

I'll also comment on a few other things that could be changed but don't need to.

vkaravir commented on src/messages.js:
```diff
@@ -39,6 +39,10 @@
     }
+    // Narrating every call to umsg
+    if($(this.jsav.container).attr("voice") === "true") {
+      JSAV.ext.textToSpeech(msg);
```
This could be changed to `this.jsav.textToSpeech)msg);`. When things are added to JSAV.ext, they'll become instance functions on jsav objects.

vkaravir commented on src/messages.js:
```diff
@@ -49,11 +53,14 @@
       this.jsav.container.trigger("jsav-message", [newValue, this.options]);
+      // Narrating for Backward buttons
+      if($(this.jsav.container).attr("voice") === "true") {
+        JSAV.ext.textToSpeech(newValue);
```
This can also use `this.jsav.textToSpeech`

vkaravir commented on src/messages.js:
```diff
@@ -65,9 +72,111 @@
+    modifiedMsg = modifiedMsg.replace(/\\mbox/gi, "");
+    modifiedMsg = modifiedMsg.replace(/n-/gi, "n minus");
+    modifiedMsg = modifiedMsg.replace(/m-/gi, "m minus");
+    modifiedMsg = modifiedMsg.replace(/%/gi, "remainder");
```
Some of these seem quite specific to the markup the visualizations are using. I'd love to see a way for an animation to add to this list of replacements. Not needed for merge, but just something to think about in the future as this way will not scale that well.

vkaravir commented on src/messages.js:
```diff
@@ -65,9 +72,111 @@
+    // Sound button click event
+    $elem.click(function() {
+      var txt = $(this.offsetParent).find(".jsavoutput").clone().find(".MathJax").remove().end().text();
+      JSAV.ext.textToSpeech(txt);
```
Here again could use `self.textToSpeech`

vkaravir commented on src/messages.js:
```diff
@@ -65,9 +72,111 @@
+        $(this).removeClass("sounds").addClass("soundOff");
+      });
+      //Toggle voice attribute and stop the current voice
+      var synth = window.speechSynthesis;
```
Here you could check if the API is supported by making sure synth is defined before you start using it. Or perhaps already in the beginning of the function.

vkaravir commented on src/messages.js:
```diff
@@ -65,9 +72,111 @@
+      //adding sound button if there is no sound button in the container and the container has both controls and settings button
+      if(($(this.container).parent().find(".new").length === 0) && ($(this.container).find(".jsavcontrols").length !== 0) && ($(this.container).parent().find(".jsavsettings").length !== 0)) {
+        $(this.container).parent().find(".jsavsettings").wrap("<span class='new'></span>");
+        $(this.container).parent().find(".new").append(JSAV.ext.soundSettings(self));
```
There's no need for self here, you can just use `this` (and remove the self=this from above). Also, can use `this.soundSettings()`.

vkaravir commented on src/messages.js:
```diff
@@ -65,9 +72,111 @@
+  };
+  //////////////////////////////////////////
+////////////Sound control button
+  JSAV.ext.soundSettings = function(jsav) {
```
There shouldn't be a need to pass the `jsav` object as a parameter. `soundSettings`will be an instance function of JSAV objects, and you're already using `this` to refer to the JSAV instance.

jwonderly commented:
Hi  Ville,
Thanks for the feedback! Sorry it's taken so long for me to address your comments; I've had a busy week.
 
Here's a summary of the changes I made in response to your comments:
 
- `soundSettings()` is no longer a JSAV extension; since it just creates the button element, it shouldn't be called outside the `init()` function anyway.
- I made it possible for developers to modify the set of modifications made to the text before the it is given to the speechSynthesis API. Developers can access and modify the global defaults through `JSAV_OPTIONS.narration.replacements`. These global defaults are initially set in core.js at the bottom of the file. 
 
If a developer wants to extend the default list for a single visualization, then they can do:
```javascript
var avOptions = {
    narration: {
        appendReplacements = [
            {searchValue: /\\theta/gi, replaceValue: 'theta'}
        ]
    }
};
var av = new JSAV('myAV', avOptions);
```
If a developer wants to override the default list, they can do:
```javascript
var avOptions = {
    narration: {
        overrideReplacements = [
            {searchValue: /\\theta/gi, replaceValue: 'theta'}
        ]
    }
};
var av = new JSAV('myAV', avOptions);
```

I chose to do it this way because when providing an array of length N for `avOptions.narration.replacements` , jQuery's extend method simply overwrites the first N elements in the `JSAV_OPTIONS.narration.replacements`

- A developer would now enable narration by doing `JSAV_OPTONS.narration.enabled = true`
- Added checks to ensure the browser supports the speechSynthesis API before creating the narration button and on every call to `JSAV.ext.textToSpeech()`.
- Removed the hard-coded language for the narration voice. It now uses `options.lang`. Note that since the speechSynthesis API uses IETF language codes (which includes a region specifier for languages with regional variations) I also added `JSAV_OPTIONS.narration.langMap` which defines which regional variation should be used for such languages. This mapping can be modified by the developer.
 
If you find these changes acceptable, I will update the documentation.

vkaravir commented:
Looks really good @jwonderly, thank you for making the changes! And apologies for taking so long, I thought I had already merged this.
